### DISCUSSION TITLE
Docker compose boilerplate

### DISCUSSION THREAD

**Discussion by rozeappletree** [fastapi/fastapi#7234] [Questions] [answered]
Hi,

Can anyone share working `docker-compose.yml` file for basic FastAPI code.
Preferred if connected to PostgreSQL

Thanks,
Rakesh

**Comment by IanSaucy** [comment]
You might find the `docker-compose.yml` file provided in [this project useful](https://github.com/Buuntu/fastapi-react). It's much simpler than the [full-stack setup ](https://github.com/tiangolo/full-stack-fastapi-postgresql)provided and a bit easier to understand(although it lacks much of the fancier items). 

Using that repo, if you wanted the most basic setup, I'd just use the `backend` and `postgres` services and you should be good to go!

**Comment by rozeappletree** [comment]
Hi @IanSaucy 
Thanks for your reply. I am quite new to docker if you don't mind can you please paste minimal `Dockerfile` and `docker-compose.yml` file. Hopefully I am not asking much

**Comment by IanSaucy** [comment]
It's hard to post docker files without knowing the structure of your entire project. Docker files do not include your source code(ie all the routes for fastAPI etc), instead they say where and how it should run the code you already have -- thus it needs to factor in your file structure. If you have a repo with your project or something I might be able to help. 

I would look up some basic docker and docker-compose info/tutorials. Personally, I found it useful to understand the technology a little before implementing it into a project otherwise you end up copy-pasting without an idea of what is happening. 

Maybe one of these tutorial would be helpful? Also, I really think the react or full stack example linked in the previous example would be a good starting point. 

https://medium.com/@mike.p.moritz/using-docker-compose-to-deploy-a-lightweight-python-rest-api-with-a-job-queue-37e6072a209b
https://testdriven.io/blog/fastapi-crud/

Good luck!

**Comment by rozeappletree** [comment]
Thank you @IanSaucy Links look amazing I will definitely check them out. I was trying docker for [this minimal fast-api template](https://github.com/rakesh4real/fastapi-template).
The docker files that I have already tried are in readme file. But couldn't get them working

**Comment by mickeypash** [comment, marked as answer]
Hey @rakesh4real,
Here is a version of these files

Dockerfile
- uses Python 3.8
- exposes the app on host port 8000
- starts the app

```Dockerfile
FROM python:3.8.1

ENV PYTHONUNBUFFERED 1

EXPOSE 8000
WORKDIR /app

COPY requirements ./
RUN pip install -r requirements.txt

COPY . ./

CMD uvicorn --host=0.0.0.0 app.main:app
```

docker-compose.yml
- Runs the app service
- Runs a Postgres database (but feel free to change that to something else or remove it)

```yaml
version: '3.7'

services:

  app:
    build: .
    ports:
      - "8000:8000"
    depends_on:
      - db
    environment:
      SQLALCHEMY_DATABASE_URL: "postgres://postgres:test@db/postgres"
  db:
    image: postgres:9.6
    ports:
      - "5432:5432"
    environment:
      POSTGRES_PASSWORD: test
```

Hope this helps

**Comment by rozeappletree** [comment]
Thank you @mickeypash @IanSaucy. Issue resolved closing it. I have updated the [repo](https://github.com/rakesh4real/fastapi-template) with files.

**Comment by Garito** [comment]
Giving the docker file, could you explain how PYTHONUNBUFFERED afects fastapi?
Thanks

**Comment by mickeypash** [comment]
By setting the `PYTHONUNBUFFERED=1` or `TRUE` you are telling Python to not buffer the `stdout/stderr` streams, [see docs](https://docs.python.org/3/using/cmdline.html#cmdoption-u). In other words, log messages are immediately dumped to the stream, thus receiving them in a timely manner. This means that you can do `docker logs <container-id>` and see the logs for your FastAPI app as they come in rather than having to wait for them to be flushed from the buffer.

**Comment by rozeappletree** [comment]
> "PYTHONUNBUFFERED also ensures that no partial output is held in a buffer somewhere and never written to terminal logs in case the python application crashes."

[source](https://stackoverflow.com/questions/59812009/what-is-the-use-of-pythonunbuffered-in-docker-file)

**Comment by Garito** [comment]
That will explain why prints are printed after restarting the server

Thanks!

**Comment by tiangolo** [comment]
Thanks for the help here everyone! :clap: :bow: 

Thanks for reporting back and closing the issue :+1:
